File: libpretixsync/src/main/java/eu/pretix/libpretixsync/sync/ReusableMediaSyncAdapter.java
package eu.pretix.libpretixsync.sync;

import org.joda.time.DateTime;
import org.joda.time.DateTimeZone;
import org.joda.time.Duration;
import org.joda.time.format.DateTimeFormat;
import org.joda.time.format.DateTimeFormatter;
import org.joda.time.format.ISODateTimeFormat;
import org.json.JSONArray;
import org.json.JSONException;
import org.json.JSONObject;

import java.io.IOException;
import java.io.InputStream;
import java.io.OutputStream;
import java.io.UnsupportedEncodingException;
import java.net.URLEncoder;
import java.util.ArrayList;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.List;
import java.util.Map;
import java.util.Set;
import java.util.concurrent.ExecutionException;

import eu.pretix.libpretixsync.api.ApiException;
import eu.pretix.libpretixsync.api.PretixApi;
import eu.pretix.libpretixsync.api.ResourceNotModified;
import eu.pretix.libpretixsync.db.CachedPdfImage;
import eu.pretix.libpretixsync.db.CheckIn;
import eu.pretix.libpretixsync.db.Item;
import eu.pretix.libpretixsync.db.Migrations;
import eu.pretix.libpretixsync.db.Order;
import eu.pretix.libpretixsync.db.OrderPosition;
import eu.pretix.libpretixsync.db.ResourceSyncStatus;
import eu.pretix.libpretixsync.db.ReusableMedium;
import eu.pretix.libpretixsync.utils.HashUtils;
import eu.pretix.libpretixsync.utils.JSONUtils;
import io.requery.BlockingEntityStore;
import io.requery.Persistable;
import io.requery.query.Tuple;
import io.requery.util.CloseableIterator;

public class ReusableMediaSyncAdapter extends BaseDownloadSyncAdapter<ReusableMedium, Long> {
    public ReusableMediaSyncAdapter(BlockingEntityStore<Persistable> store, FileStorage fileStorage, String eventSlug, PretixApi api, String syncCylceId, SyncManager.ProgressFeedback feedback) {
        super(store, fileStorage, "__all__", api, syncCylceId, feedback);
    }

    private String firstResponseTimestamp;
    private String lastMediumTimestamp;
    private ResourceSyncStatus rlm;

    private String rlmName() {
        return "reusablemedia";
    }

    @Override
    protected String getUrl() {
        return api.organizerResourceUrl(getResourceName());
    }

    @Override
    public void download() throws JSONException, ApiException, ExecutionException, InterruptedException {
        boolean completed = false;
        try {
            super.download();
            completed = true;
        } finally {
            ResourceSyncStatus resourceSyncStatus = store.select(ResourceSyncStatus.class)
                    .where(ResourceSyncStatus.RESOURCE.eq(rlmName()))
                    .and(ResourceSyncStatus.EVENT_SLUG.eq("__all__"))
                    .limit(1)
                    .get().firstOrNull();

            // We need to cache the response timestamp of the *first* page in the result set to make
            // sure we don't miss anything between this and the next run.
            //
            // If the download failed, completed will be false. In case this was a full fetch
            // (i.e. no timestamp was stored beforehand) we will still store the timestamp to be
            // able to continue properly.
            if (firstResponseTimestamp != null) {
                if (resourceSyncStatus == null) {
                    resourceSyncStatus = new ResourceSyncStatus();
                    resourceSyncStatus.setResource(rlmName());
                    resourceSyncStatus.setEvent_slug("__all__");
                    if (completed) {
                        resourceSyncStatus.setStatus("complete");
                    } else {
                        resourceSyncStatus.setStatus("incomplete:" + lastMediumTimestamp);
                    }
                    resourceSyncStatus.setLast_modified(firstResponseTimestamp);
                    store.upsert(resourceSyncStatus);
                } else {
                    if (completed) {
                        resourceSyncStatus.setLast_modified(firstResponseTimestamp);
                        store.upsert(resourceSyncStatus);
                    }
                }
            } else if (completed && resourceSyncStatus != null) {
                resourceSyncStatus.setStatus("complete");
                store.update(resourceSyncStatus);
            } else if (!completed && lastMediumTimestamp != null && resourceSyncStatus != null) {
                resourceSyncStatus.setStatus("incomplete:" + lastMediumTimestamp);
                store.update(resourceSyncStatus);
            }
            lastMediumTimestamp = null;
            firstResponseTimestamp = null;
        }
    }

    @Override
    public void updateObject(ReusableMedium obj, JSONObject jsonobj) throws JSONException {
        obj.setServer_id(jsonobj.getLong("id"));
        obj.setType(jsonobj.getString("type"));
        obj.setIdentifier(jsonobj.getString("identifier"));
        obj.setActive(jsonobj.getBoolean("active"));
        obj.setExpires(jsonobj.optString("expires"));
        obj.setCustomer_id(jsonobj.optLong("customer"));
        obj.setLinked_giftcard_id(jsonobj.optLong("linked_giftcard"));
        obj.setLinked_orderposition_id(jsonobj.optLong("linked_orderposition"));
        obj.setJson_data(jsonobj.toString());
    }

    @Override
    protected boolean deleteUnseen() {
        return false;
    }

    @Override
    protected JSONObject downloadPage(String url, boolean isFirstPage) throws ApiException, ResourceNotModified {
        if (isFirstPage) {
            rlm = store.select(ResourceSyncStatus.class)
                    .where(ResourceSyncStatus.RESOURCE.eq(rlmName()))
                    .and(ResourceSyncStatus.EVENT_SLUG.eq("__all__"))
                    .limit(1)
                    .get().firstOrNull();
        }
        boolean is_continued_fetch = false;

        if (rlm != null) {
            // This resource has been fetched before.
            if (rlm.getStatus() != null && rlm.getStatus().startsWith("incomplete:")) {
                // Continuing an interrupted fetch

                // Ordering is crucial here: Only because the server returns the orders in the
                // order of creation we can be sure that we don't miss orders created in between our
                // paginated requests.
                is_continued_fetch = true;
                try {
                    if (!url.contains("created_since")) {
                        url += "?ordering=datetime&created_since=" + URLEncoder.encode(rlm.getStatus().substring(11), "UTF-8");
                    }
                } catch (UnsupportedEncodingException e) {
                    e.printStackTrace();
                }
            } else {
                // Diff to last time

                // Ordering is crucial here: Only because the server returns the media in the
                // order of modification we can be sure that we don't miss media created in between our
                // paginated requests. If a medium were to be modified between our fetch of page 1
                // and 2 that originally wasn't part of the result set, we won't see it (as it will
                // be inserted on page 1), but we'll see it the next time, and we will se some
                // duplicates on page 2, but we don't care. The important part is that nothing gets
                // lost "between the pages". If a medium of page 2 gets modified and moves to page
                // one while we fetch page 2, again, we won't see it and we'll see some duplicates,
                // but the next sync will fix it since we always fetch our diff compared to the time
                // of the first page.
                try {
                    if (!url.contains("updated_since")) {
                        url += "?ordering=-updated&updated_since=" + URLEncoder.encode(rlm.getLast_modified(), "UTF-8");
                    }
                } catch (UnsupportedEncodingException e) {
                    e.printStackTrace();
                }
            }
        }

        PretixApi.ApiResponse apiResponse = api.fetchResource(url);
        if (isFirstPage && !is_continued_fetch) {
            firstResponseTimestamp = apiResponse.getResponse().header("X-Page-Generated");
        }
        JSONObject d = apiResponse.getData();
        if (apiResponse.getResponse().code() == 200) {
            try {
                JSONArray res = d.getJSONArray("results");
                if (res.length() > 0) {
                    lastMediumTimestamp = res.getJSONObject(res.length() - 1).getString("created");
                }
            } catch (JSONException e) {
                e.printStackTrace();
            }
        }
        return d;
    }

    @Override
    public CloseableIterator<ReusableMedium> runBatch(List<Long> ids) {
        return store.select(ReusableMedium.class)
                .where(ReusableMedium.SERVER_ID.in(ids))
                .get().iterator();
    }

    @Override
    CloseableIterator<Tuple> getKnownIDsIterator() {
        return store.select(ReusableMedium.SERVER_ID)
                .get().iterator();
    }

    @Override
    String getResourceName() {
        return "reusablemedia";
    }

    @Override
    Long getId(JSONObject obj) throws JSONException {
        return obj.getLong("id");
    }

    @Override
    Long getId(ReusableMedium obj) {
        return obj.getServer_id();
    }

    @Override
    ReusableMedium newEmptyObject() {
        return new ReusableMedium();
    }
}

File: libpretixsync/src/main/java/eu/pretix/libpretixsync/sync/ReusableMediaSyncAdapter.kt
package eu.pretix.libpretixsync.sync

import app.cash.sqldelight.TransactionWithoutReturn
import app.cash.sqldelight.db.QueryResult
import eu.pretix.libpretixsync.api.ApiException
import eu.pretix.libpretixsync.api.PretixApi
import eu.pretix.libpretixsync.api.ResourceNotModified
import eu.pretix.libpretixsync.sqldelight.ResourceSyncStatus
import eu.pretix.libpretixsync.sqldelight.ReusableMedium
import eu.pretix.libpretixsync.sqldelight.SyncDatabase
import eu.pretix.libpretixsync.sync.SyncManager.ProgressFeedback
import org.json.JSONException
import org.json.JSONObject
import java.io.UnsupportedEncodingException
import java.net.URLEncoder
import java.util.concurrent.ExecutionException

class ReusableMediaSyncAdapter(
    db: SyncDatabase,
    fileStorage: FileStorage,
    api: PretixApi,
    syncCycleId: String,
    feedback: ProgressFeedback?,
) : BaseDownloadSyncAdapter<ReusableMedium, Long>(
    db = db,
    api = api,
    syncCycleId = syncCycleId,
    eventSlug = "__all__",
    fileStorage = fileStorage,
    feedback = feedback,
) {

    private var firstResponseTimestamp: String? = null
    private var lastMediumTimestamp: String? = null
    private var rlm: ResourceSyncStatus? = null

    override fun getResourceName(): String = "reusablemedia"

    private fun rlmName(): String = "reusablemedia"

    override fun getUrl(): String = api.organizerResourceUrl(getResourceName())

    override fun getId(obj: ReusableMedium): Long = obj.server_id!!

    override fun getId(obj: JSONObject): Long = obj.getLong("id")

    override fun getJSON(obj: ReusableMedium): JSONObject = JSONObject(obj.json_data!!)

    override fun queryKnownIDs(): MutableSet<Long> {
        val res = mutableSetOf<Long>()
        db.reusableMediumQueries.selectServerIds().execute { cursor ->
            while (cursor.next().value) {
                val id = cursor.getLong(0)
                    ?: throw RuntimeException("server_id column not available")

                res.add(id)
            }
            QueryResult.Unit
        }

        return res
    }

    override fun insert(jsonobj: JSONObject) {
        db.reusableMediumQueries.insert(
            active = jsonobj.getBoolean("active"),
            customer_id = jsonobj.optLong("customer"),
            expires = jsonobj.optString("expires"),
            identifier = jsonobj.getString("identifier"),
            json_data = jsonobj.toString(),
            linked_giftcard_id = jsonobj.optLong("linked_giftcard"),
            linked_orderposition_id = jsonobj.optLong("linked_orderposition"),
            server_id = jsonobj.getLong("id"),
            type = jsonobj.getString("type"),
        )
    }

    override fun update(obj: ReusableMedium, jsonobj: JSONObject) {
        db.reusableMediumQueries.updateFromJson(
            active = jsonobj.getBoolean("active"),
            customer_id = jsonobj.optLong("customer"),
            expires = jsonobj.optString("expires"),
            identifier = jsonobj.getString("identifier"),
            json_data = jsonobj.toString(),
            linked_giftcard_id = jsonobj.optLong("linked_giftcard"),
            linked_orderposition_id = jsonobj.optLong("linked_orderposition"),
            type = jsonobj.getString("type"),
            id = obj.id,
        )
    }

    override fun delete(key: Long) {
        db.reusableMediumQueries.deleteByServerId(key)
    }

    override fun deleteUnseen(): Boolean {
        return false
    }

    override fun runInTransaction(body: TransactionWithoutReturn.() -> Unit) {
        db.reusableMediumQueries.transaction(false, body)
    }

    override fun runBatch(parameterBatch: List<Long>): List<ReusableMedium> =
        db.reusableMediumQueries.selectByServerIdList(parameterBatch).executeAsList()

    @Throws(
        JSONException::class,
        ApiException::class,
        ExecutionException::class,
        InterruptedException::class
    )
    override fun download() {
        var completed = false
        try {
            super.download()
            completed = true
        } finally {
            val resourceSyncStatus = db.resourceSyncStatusQueries.selectByResourceAndEventSlug(
                resource = rlmName(),
                event_slug = "__all__",
            ).executeAsOneOrNull()

            // We need to cache the response timestamp of the *first* page in the result set to make
            // sure we don't miss anything between this and the next run.
            //
            // If the download failed, completed will be false. In case this was a full fetch
            // (i.e. no timestamp was stored beforehand) we will still store the timestamp to be
            // able to continue properly.
            if (firstResponseTimestamp != null) {
                if (resourceSyncStatus == null) {
                    val status = if (completed) {
                        "complete"
                    } else {
                        "incomplete:$lastMediumTimestamp"
                    }

                    db.resourceSyncStatusQueries.insert(
                        event_slug = "__all__",
                        last_modified = firstResponseTimestamp,
                        meta = null,
                        resource = rlmName(),
                        status = status,
                    )

                } else {
                    if (completed) {
                        db.resourceSyncStatusQueries.updateLastModified(
                            last_modified = firstResponseTimestamp,
                            id = resourceSyncStatus.id,
                        )
                    }
                }
            } else if (completed && resourceSyncStatus != null) {
                db.resourceSyncStatusQueries.updateStatus(
                    status = "complete",
                    id = resourceSyncStatus.id,
                )
            } else if (!completed && lastMediumTimestamp != null && resourceSyncStatus != null) {
                db.resourceSyncStatusQueries.updateStatus(
                    status = "incomplete:$lastMediumTimestamp",
                    id = resourceSyncStatus.id,
                )
            }
            lastMediumTimestamp = null
            firstResponseTimestamp = null
        }
    }

    @Throws(ApiException::class, ResourceNotModified::class)
    override fun downloadPage(url: String, isFirstPage: Boolean): JSONObject? {
        if (isFirstPage) {
            rlm = db.resourceSyncStatusQueries.selectByResourceAndEventSlug(
                resource = rlmName(),
                event_slug = "__all__",
            ).executeAsOneOrNull()
        }
        var is_continued_fetch = false

        var resUrl = url
        rlm?.let {
            // This resource has been fetched before.
            if (it.status != null && it.status.startsWith("incomplete:")) {
                // Continuing an interrupted fetch

                // Ordering is crucial here: Only because the server returns the orders in the
                // order of creation we can be sure that we don't miss orders created in between our
                // paginated requests.

                is_continued_fetch = true
                try {
                    if (!resUrl.contains("created_since")) {
                        resUrl += "?ordering=datetime&created_since=" + URLEncoder.encode(
                            it.status.substring(11), "UTF-8"
                        )
                    }
                } catch (e: UnsupportedEncodingException) {
                    e.printStackTrace()
                }
            } else {
                // Diff to last time

                // Ordering is crucial here: Only because the server returns the media in the
                // order of modification we can be sure that we don't miss media created in between our
                // paginated requests. If a medium were to be modified between our fetch of page 1
                // and 2 that originally wasn't part of the result set, we won't see it (as it will
                // be inserted on page 1), but we'll see it the next time, and we will se some
                // duplicates on page 2, but we don't care. The important part is that nothing gets
                // lost "between the pages". If a medium of page 2 gets modified and moves to page
                // one while we fetch page 2, again, we won't see it and we'll see some duplicates,
                // but the next sync will fix it since we always fetch our diff compared to the time
                // of the first page.

                try {
                    if (!resUrl.contains("updated_since")) {
                        resUrl += "?ordering=-updated&updated_since=" + URLEncoder.encode(
                            it.last_modified,
                            "UTF-8"
                        )
                    }
                } catch (e: UnsupportedEncodingException) {
                    e.printStackTrace()
                }
            }
        }

        val apiResponse = api.fetchResource(resUrl)
        if (isFirstPage && !is_continued_fetch) {
            firstResponseTimestamp = apiResponse.response.header("X-Page-Generated")
        }
        val d = apiResponse.data
        if (apiResponse.response.code == 200) {
            try {
                val res = d!!.getJSONArray("results")
                if (res.length() > 0) {
                    lastMediumTimestamp = res.getJSONObject(res.length() - 1).getString("created")
                }
            } catch (e: JSONException) {
                e.printStackTrace()
            }
        }
        return d
    }

}
